Ms 1850 system doc (#57)
Meshcloud/meshcloud-docs: docs/meshstack.logging.md

---
id: meshstack.logging
title: Logging & Auditing
---

> The meshStack system consists of different components. On the one hand there are meshStack components, like the backend, replicators and further microservices. Additionally, Keycloak is used as an authentication and authorization system. Logs are therefore distributed among multiple components. In order to provide central access to the log files for administrators, a central logging system like an ELK (Elasticsearch, Logstash, Kibana) stack is used. It allows searching logs of multiple components.

Reviewer: You could add a link the the Keycloak website if someone wants to get more info about it

Besides log files, event tables are also used in meshStack. They can be used for auditing purposes. The events can be used to answer e.g. the following questions:

- Who created a certain project?
- Who gave or revoked a certain user access to a project?
- Who updated the payment information on a project?

> The event tables always contain a reference (DB Id) to the user who created it and the timestamp of the event.

Reviewer: DB Id... I dont think it is clear to everyone what it means

Author: ok, i removed it, it is not relevant here

| Edit Project (Assign/Unassign users, edit user roles, edit billing information) | ProjectEvent table, mesh.log |

## Logging of business-related administrative access

Administrators and meshPartner have access rights, that exceed the normal user’s capabilities. Therefore administrative actions require special control and traceability.

| Event (What?) | Log-File / DB-Table (Where) |
| --------------------- | --- |
| Assign yourself to a customer | User Event table, Customer Event Table, mesh.log |
| Change project quota | Project Event table, mesh.log |
| Update financial information of a project | Project Event table, mesh.log |
| Send message to customer | mesh.log |
| Change status of customer (Disabled/Verified) | Customer Event table, mesh.log |
| Deletion of a user | User table in deletedOn and deletedBy fields |

## Security relevant events

This section overlaps in some parts with the previously mentioned log files, but it summarizes all logs that are written for security related actions, like giving and revoking access.

| Event (What?) | Log-File / DB-Table (Where) |
| --------------------- | --- |

> | Successful and Denied Login attempts, as well as logouts | Keycloak Events, events.log |

Reviewer: ```suggestion
| Successful and denied login attempts, as well as logouts | Keycloak Events, events.log |
```

| Create, change, lock, unlock and delete accounts and roles | Customer/Project Event table, mesh.log |
| Password changes - Authorization via meshIdB | Keycloak Events, events.log, mesh.log |
| Password changes - Authorization via federated IdP | Federated IdP |
| Access Right changes (i.e. user rights) | Customer/Project Event table, mesh.log |
| Changes to logging configuration (especially deactivation of logging) | Can only be done by an operator, no logging atm |
| Start and stop administrative processes (Batch-Jobs) | mesh.log |

## Structure / Content, Format and Retention

### Structure

All log file entries always contain a formatted UTC timestamp and the log level.

### Keycloak Log

> The keycloak log only contains operational keycloak information. It is mainly used for error analysis. Therefore only a few fields are provided in this log file.

Reviewer: ```suggestion
The Keycloak log only contains operational information. It is mainly used for error analysis. Therefore only a few fields are provided in this log file.
```

| Log-File / DB Entry | Minimum Retention Time | Maximum Retention Time |
| --- | --- | --- |
| Keycloak Events in DB | 2 month | 12 month |
| Keycloak Events in log file | 1 month | 6 month |
| Keycloak Logs | 1 month | 6 month |
| meshStack component logs | 1 month | 6 month |
| meshStack Events in DB | 12 month | 5 years |

## Deletion Process of log files

As all components use logging frameworks, these frameworks are configured to automatically delete log files after a configured period of time.

The retention time of Keycloak Events in the Keycloak Database is configured within Keycloak. It is a configuration option of the tool.

meshStack events are deleted by a weekly job, that checks, whether events exceeded the defined lifetime.

As the database is also backed up, deleted events will persist in the backup, as long as the backup exists. Database backups of meshStack and Keycloak are persisted for a configured amount of time (e.g. 24 days).

## Location of Log Files

> All log files are stored locally in the VM, container or PaaS system, where the component is running. These instances require authentication (e.g. via SSH with a private key) for access to the logs.

Reviewer: ```suggestion
All log files are stored locally in the VM, container or PaaS system, where the component is running. For these instances user authentication  is required (e.g. via SSH with a private key) to access their logs.
```
